### ISSUE TITLE
Support for universal binaries

### ISSUE THREAD

**Issue by bclothier** [WaynePhillipsEA/twinbasic#515]
I believe that twinBasic will have lot of potential for use with Office or other VBA hosts. As you may know, VBA codebase can run in both 32-bit and 64-bit mode. It would be much more convenient to be able to ship a DLL with a Office application that would run in either mode. While one can create an installer which can then install the proper version, nothing can beat the simplicity of a xcopy development. 

Therefore, my wish is that it would be possible to build an output that has both 32-bit and 64-bit binaries embedded and ship with some kind of bootstrapper so that when it is used, the correct bitness will be used. That would then allow us to ship a single executable file and gain more wide adoption due to not needing an installer. That will be valuable especially for locked-down environments.

For general reference: [universal binaries](https://en.wikipedia.org/wiki/Universal_binary) and [for Windows specifically](https://en.wikipedia.org/wiki/Fat_binary#Fatpack)

**Comment by WaynePhillipsEA** [comment]
The PE file format itself doesn't support nativecode fat binaries at all; I wish it did!.  The closest thing is .NET executables compiled with 'Any CPU', where the OS loads the appropriate bitness of the CLR, and the code is then compiled from the IL-code into nativecode at runtime etc.   

As I understand, this special support for the CLR loading is baked into the OS, and is exclusively for .NET executables.  So the only option would seem to be for the tB compiler to output .NET IL-code... which of course is outside the scope of the project.

This was one of the main pain points that my Virtual-COM technology solves.  Virtual-COM, as seen in vbWatchdog, integrates the nativecode binaries into the host VBA project, meaning there are no DLLs needing to be distributed to end-users at all and the correct nativecode bitness is picked at runtime automatically.  FWIW, I'd love to see Virtual-COM become integrated into twinBASIC at some point, as it solves so many distribution problems.

**Comment by bclothier** [comment]
Just making sure, did you look at [fatpack](https://github.com/sjmulder/fatpack) already? 

I imagine this solves the problem with executables but not necessarily the loading of the ActiveX DLL. I suspect that a bootstrapper will be necessary. Even in the case of a .NET DLL, I have to use your vbDotNetLoader to avoid the need to registry COM objects and thus achieve a truly COM-free registration for VBA projects. That's on the account of us needing to generate a specific type library for the bitness and registry that.

**Comment by WaynePhillipsEA** [comment]
> Just making sure, did you look at [fatpack](https://github.com/sjmulder/fatpack) already?

Yes I did.  It's basically a Win32 executable that extracts the real Win32 or Win64 EXE to a temp file and executes from there.  It's a workaround, but certainly nothing like a real universal / fat binary.    

And yes it's not an option for DLLs... they have to be of the correct bitness (...or be a special .NET AnyCPU build) else LoadLibrary will simply fail.

**Comment by WaynePhillipsEA** [comment]
Although not ideal, there's also the [DllSurrogate method](https://docs.microsoft.com/en-us/windows/win32/com/dllsurrogate) of making an ActiveX/COM library work when the bitness doesn't match.  Of course this method causes the library to be run out-of-process (with automatic IPC that marshals the COM calls), and as such is no real substitute for a normal, in-process library.

**Comment by bclothier** [comment]
Thanks for the background. Based on that information, I think your original idea of using V-COM as a generic bootstrapper to select and load the correct native code binaries is probably the best way and makes it much more convenient to handle compared to reg-free COM which involves manifest files for the executable (e.g. `MSACCESS.EXE`, not `Widgets.accdb`). If the V-COM can also support late-bound event handling similar to what your `vbDotNetLoader` does, then it would be possible to ship it late-bound without any downsides and not have to create a VBA module for all the signatures.

**Comment by wqweto** [comment]
If I might chime in: can we get all this "V-COM" goodness optionally i.e. for the rest of us to be able to produce "uncomfortble" regular x86/x64 ActiveX DLLs and use these reg-free in use-cases outside MS Access?
